### PR TITLE
reindex broken search data

### PR DESCRIPTION
**Pull request by fantasticfears**
Since `cppjieba` is in place, it changes the search data vector. Basically magically rebuilt the search index.

@SamSaffron concerns for performance hit?

### CODE REVIEW

File: app/services/search_indexer.rb [hunk 1]
```diff
@@ -149,4 +225,18 @@ def characters(string)
...
+    if SiteSetting.search_index_core_version < 1
+      if ['zh_TW', 'zh_CN', 'ja', 'ko'].include?(SiteSetting.default_locale) || SiteSetting.search_tokenize_chinese_japanese_korean
+        self.force_reindex_posts_topics
+        self.force_reindex_categories
+        self.force_reindex_users
+      end
+
+      SiteSetting.search_index_core_version = 1
+    end
+  end
```
Comment (SamSaffron) [inline app/services/search_indexer.rb]: this should be in a once off job, I don't particularly like having this in the core class.

version should be stored in the search table and then we can have a regular weekly job that forces are reindex on posts that have an earlier version and cleanly recover from partial re-index

note I am going to be out for a week, so it is going to take a while to merge this in.

Comment (fantasticfears) [inline app/services/search_indexer.rb]: Sure, got it

File: db/migrate/20170803123704_add_version_to_search_data.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class AddVersionToSearchData < ActiveRecord::Migration
+  def change
+    add_column :post_search_data, :core_version, :integer, default: 0
```
Comment (SamSaffron) [inline db/migrate/20170803123704_add_version_to_search_data.rb]: hmmm ... lets use the name `version` here, core version is a bit confusing

**Comment by SamSaffron** [comment]
lets remove the once off and instead default to version 0 and have version 1 be in code. 

I think we are well overdue for a full search re-index due to other bugs and this helps remove a whole bunch of code.

**Comment by fantasticfears** [comment]
Done

File: app/jobs/scheduled/reindex_search.rb [hunk 3]
```diff
@@ -1,10 +1,69 @@
...
+
 module Jobs
   # if locale changes or search algorithm changes we may want to reindex stuff
   class ReindexSearch < Jobs::Scheduled
     every 1.day
 
     def execute(args)
-      Search.rebuild_problem_posts
+      limiter = Search::QueryLimiter.new(args[:limit] || 10_000)
+      limiter.step { |limit| rebuild_problem_categories(limit) }
```
Comment (SamSaffron) [inline app/jobs/scheduled/reindex_search.rb]: I don't really think we need all this fancy...  just limit to 10000 topics / 10000 posts / 10000 users. no need to do a limit on 10000 things.  code will be much easier to understand, one less class.

File: spec/services/search_indexer_spec.rb [hunk 4]
```diff
@@ -13,14 +13,16 @@
     expect(raw_data.split(' ').length).to eq(2)
   end
 
+  let(:version_num) { 12345 }
   it 'correctly indexes a post' do
     data = "<a>This</a> is a test"
-
+    stub_constant(Search, 'INDEX_VERSION', version_num)
```
Comment (SamSaffron) [inline spec/services/search_indexer_spec.rb]: mocking and stubbing makes me super uneasy. in this context. 

I would ... index something, then update the row to -1 and make sure it gets re-indexed, something like that ... no need for fancy mocking.

**Comment by fantasticfears** [comment]
Fixed

**Comment by SamSaffron** [comment]
now we are talking! looking great. 

I am considering adding an index on version,locale,(post_id|topic_id) so the left join is fast in the job. but since the query only runs daily this is not critical. worth measuring the impact though.

**Comment by SamSaffron** [comment]
Oh ... one thing I noticed... we are loading 10k posts / topics etc into memory... that is not ideal. 

What we should be doing is selecting ids and then using find_each or something along those lines so the in-memory batches are not huge and we dont bloat sidekiq

**Comment by SamSaffron** [comment]
ok committed a fix for memory bloating.

**Comment by fantasticfears** [comment]
Thanks!

**Pull request merged**